File: heron/api/src/java/com/twitter/heron/dsl/Builder.java
//  Copyright 2017 Twitter. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package com.twitter.heron.dsl;

import com.twitter.heron.dsl.impl.BuilderImpl;

/**
 * Builder is used to register all sources. Builder thus keeps track
 * of all the starting points of the computation dag and uses this
 * information to build the topology
 */
public interface Builder {
  static Builder createBuilder() {
    return new BuilderImpl();
  }

  /**
   * All sources of the computation should register using addSource.
   * @param supplier The supplier function that is used to create the streamlet
   */
  <R> Streamlet<R> newSource(SerializableSupplier<R> supplier);

  /**
   * Same as above except returns KVStreamlet
   * @param supplier
   * @param <K>
   * @param <V>
   * @return
   */
  <K, V> KVStreamlet<K, V> newKVSource(SerializableSupplier<KeyValue<K, V>> supplier);

  /**
   * Creates a new Streamlet using the underlying generator
   * @param generator The generator that generates the tuples of the streamlet
   * @param <R>
   * @return
   */
  <R> Streamlet<R> newSource(SerializableGenerator<R> generator);

  /**
   * Same as above except returns a KVStreamlet
   * @param generator The genertor that generates the tuples of the streamlet
   * @param <K>
   * @param <V>
   * @return
   */
  <K, V> KVStreamlet<K, V> newKVSource(SerializableGenerator<KeyValue<K, V>> generator);
}

File: heronpy/streamlet/builder.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

'''builder.py: module for creating streamlets'''

from heronpy.streamlet.generator import Generator
from heronpy.streamlet.impl.supplierspout import SupplierStreamlet
from heronpy.streamlet.impl.generatorspout import GeneratorStreamlet

class Builder:
  """A Builder object is used to build the functional API DAG in Heron."""
  def __init__(self):
    """
    """
    self._sources = []

  def new_source(self, source):
    """Adds a new source to the computation DAG"""

    source_streamlet = None
    if callable(source):
      source_streamlet = SupplierStreamlet(source)
    elif isinstance(source, Generator):
      source_streamlet = GeneratorStreamlet(source)
    else:
      raise RuntimeError("Builder's new source has to be either a Generator or a function")

    self._sources.append(source_streamlet)
    return source_streamlet

  # pylint: disable=protected-access
  def build(self, bldr):
    """Builds the topology and returns the builder"""
    stage_names = set()
    for source in self._sources:
      source._build(bldr, stage_names)
    for source in self._sources:
      if not source._all_built():
        raise RuntimeError("Topology cannot be fully built! Are all sources added?")
